# Question 1

In [5]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [6]:
import pandas as pd

In [7]:
df = pd.read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv')

In [8]:
missing_values_per_column = df.isnull().sum()

In [11]:
print("Missing values per column:")
print(missing_values_per_column)

Missing values per column:
row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


In [12]:
total_missing_values = df.isnull().sum().sum()
print(f"\nTotal number of missing values in the dataset: {total_missing_values}")


Total number of missing values in the dataset: 12


ChatBot link: https://chatgpt.com/share/2a6c5a16-f4a4-4fad-83e4-9c5ffea1a0b0

# Question 2

In [3]:
import pandas as pd

In [4]:
file_path = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'

In [5]:
data = pd.read_csv(file_path)

In [6]:
rows, columns = data.shape

In [7]:
print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 391
Number of columns: 11


Observation - a singular data point within a given dataset. An observation represents a specific moment in which data is gathered.
Variables - a specific attribute/characteristic with differing properties between observations. Variables represent a measurable/recordable property for each observation in a given dataset

ChatBot link: https://chatgpt.com/share/1f219df1-9b23-4e83-b685-eb0fba645b48

# Question 3

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv')

In [3]:
print("First 5 rows of the dataset:")
print(df.head())

First 5 rows of the dataset:
   row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/... 

In [4]:
print("\nDataset Information:")
print(df.info())


Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB
None


In [5]:
print("\nStatistical Summary:")
print(df.describe())


Statistical Summary:
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000


In [6]:
print("\nMissing Values per Column:")
print(df.isnull().sum())


Missing Values per Column:
row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


In [10]:
print("Available columns in the dataset:")
print(df.columns)

Available columns in the dataset:
Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')


In [11]:
print("\nCategorical Column Summary for 'Birthday':")
print(df['birthday'].value_counts())


Categorical Column Summary for 'Birthday':
birthday
1-27     2
12-5     2
7-31     2
3-26     2
8-3      2
        ..
4-3      1
10-26    1
7-23     1
12-8     1
3-8      1
Name: count, Length: 361, dtype: int64


ChatBot link: https://chatgpt.com/share/56d6f449-743e-467a-8d02-0dcbd4c5dde9

# Question 4

In [12]:
import pandas as pd

In [13]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

In [14]:
print("Shape of the dataset:", df.shape)

Shape of the dataset: (891, 15)


In [15]:
print("Descriptive statistics:")
print(df.describe())

Descriptive statistics:
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


Discrepency - based on df.shape, this dataset has 15 columns, yet only 6 are presented in df.describe. Reasoning - describe function only shows columns with numerical values. To fix this, use the following command:

In [17]:
df.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

Discrepency - based on df.shape, this dataset has 891 rows, and this is reflected in most columns within the dataset, but not all. For example, there are only 714 rows listed in the age column, meaning 177 values are missing (891 - 714 = 177). There are also discrepancies in the number of rows for the columns embarked, deck, and embark_town.

To conclude, major discrepancy is this: in df.shape, we get total number of rows without missing values and total number of columns no matter the type of value, whereas in df.describe() they only show columns with int/float values and in the count row, only represent the amount of non-missing values.

ChatBot link: https://chatgpt.com/share/d02038d2-d6d3-442e-9f52-c7fe52e31da6

# Question 5

Attributes are considered as characteristics or properties of an object, where certain information regarding the properties of the data are stored, such as dimensions. There are no parentheses because they cannot be called by the program, their only purpose is to store and display values to the user when accessed. On the other hand, methods are functions associated with the object that can be called by the program, being able to perform specific actions on the object, hence why it requires parentheses to perform these operations.

ChatBot link: https://chatgpt.com/share/7964cf1f-e306-4734-99eb-9e5d6ee5cda1

# Question 6

Summary Statistics:
Count - provides the number of non-missing values in the column with type float.
Mean - avg value of a column (adding every value and dividing by the number of values) *remembering that df.describe() represents numerical data only unless an argument is provided otherwise*
Standard Deviation - this measures the spread of the provided values around the mean, calculating the avg distance from the mean.
Min - smallest value in the column.
25% - also known as the 1st quartile (Q1), is the value below which 25% of the data points are found when put in increasing order, allowing us to better understand the distribution of these lower parts of the data.
50% - also known as the 2nd quartile (Q2), the middle value of the given dataset when in increasing order, with 50% of the data above and below this quartile. Represents the central tendency (value that represents the center of the data distribution) without being altered by outliers.
75% - also known as the 3rd quartile (Q3), is the value below which 75% of the data points are found when put in increasing order, allowing us to better understand the distribution of these upper portions of the data.
Max - biggest value in the column.

ChatBot link: https://chatgpt.com/share/369d0526-e0c1-45d9-8a18-157d770def93

# Question 7

An instance "use case" where df.dropna() would be preferred over the use of del df['col'] is when the amount of missing data is very limited and you want the dataset to be preserved as much as possible, thus the decision to drop the missing rows of a dataset allows the size of the dataset to still be large as opposed to using del df['col']. Specific Example - 10 missing rows in a 1000 row dataset retains 990 complete rows

An "opposite use case" for using del df['col'] would be when entire columns have too much missing data in the dataset for the information to be valuable and useful. If a column has 75% of data missing, choosing to delete that whole column using del df['col'] could prove more efficient as opposed to dealing with the missing values.

When using both df.dropna() and del df['col'] together, applying del df['col'] could be important in a situation where one column has a lot of missing data, which is displayed in a more scattered way across rows, thus making the application of del df['col'] more efficient for clearing out this missing data and satisfying the goal of retaining as much data as possible.

In [2]:
import pandas as pd

In [18]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

In [19]:
print("Before:")
(df.describe(include='all'))

Before:


survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

In [21]:
del df['deck']

In [22]:
print("After:")
(df.describe(include='all'))

After:


survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male  embark_town alive alone  
count   891.000000      889    891  891        891          889   891   891  
unique         NaN        3      3    3          2            3     2     2  
top            NaN        S  Third  man       True  Southampton    no  True  
freq           NaN      644    491  537        537          644   549   537  
mean     32.204208      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
std      49.693429      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
min       0.000000      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
25%       7.910400      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
50%      14.454200      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
75%      31.000000      NaN    NaN  NaN        NaN          NaN   NaN   NaN  
max     512.329200      NaN    NaN  NaN        NaN          NaN   NaN   NaN

Beforehand the dataset consisted of a column with a lot of missing values, that being the 'deck' column. By getting rid of this column in the dataset, we reduce the amount of missing data and improve the overall quality of the dataset, potentially making the analysis more accurate and efficient. Afterwards we can see a more well-represented set of data.

ChatBot link: https://chatgpt.com/share/66e39fc9-e0b8-8012-af60-ce928b0c308d

# Question 8

In [23]:
import pandas as pd

In [24]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

In [30]:
grouped_stats = df.groupby("who")["fare"].describe()
grouped_stats

count       mean        std    min      25%    50%       75%       max
who                                                                          
child   83.0  32.785795  33.466438  7.225  15.9000  26.25  32.19375  211.3375
man    537.0  24.864182  44.021339  0.000   7.8542   9.50  26.38750  512.3292
woman  271.0  46.570711  60.318995  6.750  10.5000  23.25  65.00000  512.3292

The code df.groupby("col1")["col2"].describe() groups data based on certain columns, generating descriptive statistics for each group.

df.describe() captures the entire dataset, including missing values, and so the count value would differ from df.groupby("col1")["col2"].describe() where only non-null values are displayed. Additionally, df.groupby("col1")["col2"].describe() only looks at subsets of the dataframe grouped by col1, whereas df.describe() looks at the entire dataframe.

In [1]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

NameError: name 'pd' is not defined

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

HTTPError: HTTP Error 404: Not Found

In [4]:
df = pd.read_csv('https://raw.githobusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

URLError: <urlopen error [Errno -2] Name or service not known>

In [5]:
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

In [6]:
grouped_stats = DF.groupby("who")["fare"].describe()

NameError: name 'DF' is not defined

In [7]:
df.describe(include='all'

SyntaxError: incomplete input (47798001.py, line 1)

In [8]:
df.groupby("who")["fare"].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

In [9]:
grouped_stats = df.groupby("who")["Sex"].describe()

KeyError: 'Column not found: Sex'

In [10]:
grouped_stats = df.groupby(who)["fare"].describe()

NameError: name 'who' is not defined

Overall, I found that ChatGPT provided solutions to these errors with immediate effect, negating the use of a google search. Should I run into more complex errors in the future, if ChatGPT struggles with correcting them I can always give it more prompts to help, but for these errors it was able to identify the problem immediately.

ChatBot link: https://chatgpt.com/share/66e3a86c-a30c-8012-a549-a2565ac34d6f

# Question 9

Mostly.